File: python/p003.py
import math

"""Get the largest prime factor of a odd natural number N"""

N = 600851475143

n = N
end = math.floor(math.sqrt(n))
k = 3

while k <= end:
    if n % k == 0:
        n = n // k
        end = math.floor(math.sqrt(n))
    k += 2  # skip even numbers

print(n)

File: typescript/src/p003.ts
const N = 600851475143;

/** Get largest prime factor of n */
function LPF(n: number): number {
  let end = Math.sqrt(n);
  let k = 3;

  while (k <= end) {
    if (n % k === 0) {
      n /= k; // devise if k is a factor
      end = Math.sqrt(n); // recalculate end
    }
    k += 2; // increment k, skip even numbers
  }

  return n;
}

console.log(LPF(N));
